### PR TITLE
add catalogue list catalogue item property type #220

### LINKED ISSUE #220: Add additional catalogue item properties types
- Currently a catalogue item property can be number, boolean and Text.   
- we need to expand this list two include number range and list of numbers and list of Text and maybe Date and list of Dates
- This additional cha nge we require UI changes to allow user to input list and ranges 

For MVP only implement list type : 

- [ ] list of number 
- [ ] list of string 

- Implement the allowed values list in the catalogue categories dialog
- #276 
- Implement the allowed values list in the items and catalogue items dialogues 
- [ ] #292

### PR DESCRIPTION
**Pull request by joshuadkitenge**
## Description
add the list type to the catalogue item properties 

## Testing instructions

Add a set up instructions describing how the reviewer should test the code

- [ ] Review code
- [ ] Check Actions build
- [ ] Review changes to test coverage
- [ ] test the error handling

## Agile board tracking

closes #220

### CODE REVIEW

**Review by joelvdavies — COMMENTED** [review]
Some minor things for the moment, haven't gone through all yet, but worked well when I tested it.

File: src/catalogue/category/catalogueCategoryDialog.component.tsx [hunk 1]
```diff
@@ -141,67 +139,193 @@ const CatalogueCategoryDialog = React.memo(
       setFormError(undefined);
     };
 
-    const validateFormFields = React.useCallback(() => {
-      const errorIndexes = [];
+    const validateCatalogueItems = (
```
Comment (joshuadkitenge) [inline src/catalogue/category/catalogueCategoryDialog.component.tsx:142]: Give this a better name

File: src/catalogue/category/catalogueCategoryDialog.component.tsx [hunk 2]
```diff
@@ -141,67 +139,193 @@ const CatalogueCategoryDialog = React.memo(
...
+              property.type === 'number' &&
+              (isNaN(+value) || !value)
+            ) {
+              invalidNumberIndexes.push(i);
+            }
+          });
+
+          if (
+            Array.from(new Set(duplicateIndexes)).length > 0 ||
+            Array.from(new Set(invalidNumberIndexes)).length > 0 ||
```
Comment (joshuadkitenge) [inline src/catalogue/category/catalogueCategoryDialog.component.tsx:177]: remove set fromm the invalid number and missing value

File: src/catalogue/category/catalogueCategoryDialog.component.tsx [hunk 3]
```diff
@@ -141,67 +139,193 @@ const CatalogueCategoryDialog = React.memo(
...
+            Array.from(new Set(missingValueIndexes)).length > 0
+          ) {
+            // Update listItemErrors state with the error indexes
+            setListItemErrors((prev) => [
+              ...prev,
+              {
+                index,
+                valueIndex: [
+                  ...(prev[index]?.valueIndex || []),
+
```
Comment (joshuadkitenge) [inline src/catalogue/category/catalogueCategoryDialog.component.tsx:187]: Order dependent

File: src/catalogue/category/catalogueCategoryDialog.component.tsx [hunk 4]
```diff
@@ -141,67 +139,193 @@ const CatalogueCategoryDialog = React.memo(
...
+                  ...Array.from(new Set(invalidNumberIndexes)).map((i) => ({
```
Comment (joshuadkitenge) [inline src/catalogue/category/catalogueCategoryDialog.component.tsx:188]: remove the set. look above

File: src/catalogue/category/catalogueCategoryDialog.component.tsx [hunk 5]
```diff
@@ -141,67 +139,193 @@ const CatalogueCategoryDialog = React.memo(
...
-            duplicateNames.push(name);
-          } else {
-            uniqueNames.add(name);
-          }
-        }
-        if (duplicateNames.length > 0) {
-          setDuplicatePropertyError(duplicateNames);
-          hasErrors = true;
-        }
+      if (catalogueItemPropertiesErrors.length !== 0 || formFieldErrors) {
```
Comment (joshuadkitenge) [inline src/catalogue/category/catalogueCategoryDialog.component.tsx:322]: remove catalogueItemPropertiesErrors.length !== 0

File: src/catalogue/category/cataloguePropertiesForm.component.tsx [hunk 6]
```diff
@@ -1,112 +1,345 @@
...
+      values: [...(currentField.allowed_values?.values || []), ''],
+    };
+
+    updatedFormFields[index] = {
+      ...currentField,
+      allowed_values: updatedAllowedValues,
+    };
+
+    onChangeFormFields(updatedFormFields);
+
```
Comment (joshuadkitenge) [inline src/catalogue/category/cataloguePropertiesForm.component.tsx:176]: add comment (this clears "Please create a valid list item")

File: src/catalogue/category/cataloguePropertiesForm.component.tsx [hunk 7]
```diff
@@ -1,112 +1,345 @@
...
+      onChangeListItemErrors(
+        updatedListItemErrors.filter(
+          (item) => (item.valueIndex?.length ?? 0) > 0
+        )
+      );
+    }
+  };
+
+  const isError = React.useCallback(
```
Comment (joshuadkitenge) [inline src/catalogue/category/cataloguePropertiesForm.component.tsx:274]: hasFieldError

File: src/catalogue/category/cataloguePropertiesForm.component.tsx [hunk 8]
```diff
@@ -1,112 +1,345 @@
...
+      return (
+        (atIndex?.filter((item) => {
+          return item.index === listIndex;
+        }).length ?? 0) >= 1
+      );
+    },
+    [listItemErrors]
+  );
+
+  const listErrorMessage = React.useCallback(
```
Comment (joshuadkitenge) [inline src/catalogue/category/cataloguePropertiesForm.component.tsx:326]: allowedValuesListError

File: src/catalogue/category/catalogueCategoryDialog.component.tsx [hunk 9]
```diff
@@ -141,62 +129,192 @@ const CatalogueCategoryDialog = React.memo(
...
+              (isNaN(+value) || !value)
+            ) {
+              invalidNumberIndexes.push(i);
+            }
+          });
+
+          if (
+            Array.from(new Set(duplicateIndexes)).length > 0 ||
+            // If there are more than 2 instances of the same duplicate value, it adds the index multiple times.
+            // The set removes the repeated indexes.
```
Comment (joelvdavies) [inline src/catalogue/category/catalogueCategoryDialog.component.tsx:168]: This comment should be above the line above

File: src/catalogue/category/cataloguePropertiesForm.component.test.tsx [hunk 10]
```diff
@@ -254,39 +312,712 @@ describe('Catalogue Properties Form', () => {
...
+        name: 'raduis 2',
+        allowed_values: { type: 'list', values: [''] },
+        type: 'number',
+        unit: '',
+      },
     ]);
   });
 
-  it('display error if duplicate property names are entered', async () => {
+  it('display error if duplicate property names are entered and clears when a name is change', async () => {
```
Comment (joelvdavies) [inline src/catalogue/category/cataloguePropertiesForm.component.test.tsx:332]:
```suggestion
  it('display error if duplicate property names are entered and clears when a name is changed', async () => {
```

File: src/catalogue/category/cataloguePropertiesForm.component.test.tsx [hunk 11]
```diff
@@ -254,39 +312,712 @@ describe('Catalogue Properties Form', () => {
...
+          mandatory: true,
+          name: 'Field 4',
+          type: 'string',
+          unit: '',
+        },
+      ]);
+    });
+  });
+
+  it('should add a list item when the add icon is click', async () => {
```
Comment (joelvdavies) [inline src/catalogue/category/cataloguePropertiesForm.component.test.tsx:470]:
```suggestion
  it('should add a list item when the add icon is clicked', async () => {
```

File: src/catalogue/category/cataloguePropertiesForm.component.test.tsx [hunk 12]
```diff
@@ -254,39 +312,712 @@ describe('Catalogue Properties Form', () => {
...
+          mandatory: true,
+          name: 'Field 4',
+          type: 'string',
+          unit: '',
+        },
+      ]);
+    });
+  });
+
+  it('should edit a list item when the add icon is click', async () => {
```
Comment (joelvdavies) [inline src/catalogue/category/cataloguePropertiesForm.component.test.tsx:519]:
```suggestion
  it('should edit a list item when the add icon is clicked', async () => {
```

File: src/catalogue/category/cataloguePropertiesForm.component.test.tsx [hunk 13]
```diff
@@ -254,39 +312,712 @@ describe('Catalogue Properties Form', () => {
...
+          mandatory: true,
+          name: 'Field 4',
+          type: 'string',
+          unit: '',
+        },
+      ]);
+    });
+  });
+
+  it('should error for duplicate values and incorrect type values and the error should be remove if catalogue item property is deleted', async () => {
```
Comment (joelvdavies) [inline src/catalogue/category/cataloguePropertiesForm.component.test.tsx:671]:
```suggestion
  it('should display error for duplicate values and incorrect type values and the error should be removed if the catalogue item property is deleted', async () => {
```

File: src/catalogue/category/cataloguePropertiesForm.component.test.tsx [hunk 14]
```diff
@@ -254,39 +312,712 @@ describe('Catalogue Properties Form', () => {
...
+
+    expect(onChangeAllowedValuesListErrors).toHaveBeenCalledWith([
+      {
+        index: 0,
+        errors: [{ errorMessage: 'Please enter a valid number', index: 0 }],
+      },
+    ]);
+  });
+
+  it('should error for incorrect type values and remove error if the value has been deleted', async () => {
```
Comment (joelvdavies) [inline src/catalogue/category/cataloguePropertiesForm.component.test.tsx:843]:
```suggestion
  it('should display error for incorrect type values and remove the error if the value has been deleted', async () => {
```

File: src/catalogue/category/cataloguePropertiesForm.component.test.tsx [hunk 15]
```diff
@@ -254,39 +312,712 @@ describe('Catalogue Properties Form', () => {
...
+
+    expect(onChangeAllowedValuesListErrors).toHaveBeenCalledWith([
+      {
+        index: 0,
+        errors: [{ errorMessage: 'Please enter a valid number', index: 0 }],
+      },
+    ]);
+  });
+
+  it('should error for duplicate values and incorrect type values and the error should be if the type has changed', async () => {
```
Comment (joelvdavies) [inline src/catalogue/category/cataloguePropertiesForm.component.test.tsx:910]:
```suggestion
  it('should display error for duplicate values and incorrect type values and the error should be removed if the type is changed', async () => {
```

**Review by joelvdavies — CHANGES_REQUESTED** [review]
Just a last few minor things

**Review by joelvdavies — APPROVED** [review]

**Pull request merged**